web/packages/teleport/src/Assist/context/AssistContext.tsx

```diff
@@ -0,0 +1,464 @@
+  sendMessage: (message: string) => void;
+  sendMfaChallenge: (data: WebauthnAssertionResponse) => void;
+  selectedConversationMessages: ConversationMessage[];
+  setSelectedConversationId: (conversationId: string) => void;
+}
+
+const AssistContext = createContext<AssistState & AssistContextValue>(null);
+
+let lastCommandExecutionResultId = 0;
+
+export function AssistContextProvider(props: PropsWithChildren<unknown>) {
+  const activeWebSocket = useRef<WebSocket>(null);
+  const executeCommandWebSocket = useRef<WebSocket>(null);
+
+  const { clusterId } = useStickyClusterId();
+
+  const [state, dispatch] = useReducer(reducer, {
+    conversations: {
+      loading: false,
+      data: [],
+      selectedId: null,
+    },
+    messages: {
+      loading: false,
+      streaming: false,
+      data: new Map(),
+    },
+    mfa: {
+      prompt: false,
+      publicKey: null,
+    },
+  });
+
+  async function loadConversations() {
+    dispatch({
+      type: AssistStateActionType.SetConversationsLoading,
+      loading: true,
+    });
+
+    const conversations = await service.loadConversations();
+
+    dispatch({
+      type: AssistStateActionType.ReplaceConversations,
+      conversations,
+    });
+  }
+
+  function setupWebSocket(conversationId: string) {
+    activeWebSocket.current = new WebSocket(
+      cfg.getAssistConversationWebSocketUrl(
+        getHostName(),
+        clusterId,
+        getAccessToken(),
+        conversationId
+      )
+    );
+
+    activeWebSocket.current.onmessage = async event => {
+      const data = JSON.parse(event.data) as ServerMessage;
+
+      switch (data.type) {
+        case ServerMessageType.Assist:
+          dispatch({
+            type: AssistStateActionType.AddMessage,
+            messageType: ServerMessageType.Assist,
+            message: data.payload,
+            conversationId,
+          });
+
+          break;
+
+        case ServerMessageType.AssistPartialMessageEnd:
+          dispatch({
+            type: AssistStateActionType.SetStreaming,
+            streaming: false,
+          });
+
+          break;
+
+        case ServerMessageType.AssistPartialMessage: {
+          const payload = JSON.parse(data.payload);
+
+          dispatch({
+            type: AssistStateActionType.AddPartialMessage,
+            message: payload.content,
+            conversationId,
+          });
+
+          break;
+        }
+
+        case ServerMessageType.Command: {
+          const message = await resolveServerCommandMessage(data);
+
+          dispatch({
+            type: AssistStateActionType.AddExecuteRemoteCommand,
+            message,
+            conversationId,
+          });
+
+          dispatch({
+            type: AssistStateActionType.SetStreaming,
+            streaming: false,
+          });
+
+          break;
+        }
+      }
+    };
+  }
+
+  async function createConversation() {
+    if (state.messages.streaming) {
+      dispatch({
```

Comment:
This repeats a few times. Maybe we could factor it out to some `function reset()`.

Reply:
The reset of the streaming to false? I'm not sure it's worth factoring that into another method that only dispatches